MINOR: Rolling bounce upgrade fixed broker system test

bbejeck opened the pull request:
This introduces a system test where we do a rolling bounce upgrade of streams application instances against a fixed broker version. Each broker version starting with `0.10.2` up to `trunk` is tested.  Streams applications start at `0.10.2` and do rolling upgrades up to `trunk` then do rolling downgrades back to `0.10.2`.

This PR also includes a new utility class for Streams systems tests `StreamsTest` which extends `KafkaTest` and provides some convenience methods for working with `VerifiableProducer` and `VerfiiableConsumer` and validating results in log files.

There are other system tests `streams_standby_replica_test.py` and `streams_broker_down_resilience_test.py` using common functionality that can be pulled out and extend the `StreamsTest` class instead.  I'll do that in a follow-on PR.

EDIT: Since the base class `StreamsTest` only exists in this PR so far, I've refactored the other tests as mentioned above.

I tested this by running branch builder https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1430/

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bbejeck committed:
add new kafka versions to base.sh, move functions to streams.py, add new test

bbejeck committed:
MINOR: removed changes not needed for this PR

bbejeck commented:
@dguy @guozhangwang @mjsax @vvcephei

guozhangwang commented:
@bbejeck you'll need to rebase from trunk, the jenkins failures are related to that.

guozhangwang commented on tests/kafkatest/tests/streams/streams_multiple_rolling_upgrade_test.py:
```diff
@@ -0,0 +1,114 @@
+     As new releases come out, just update the streams_upgrade_versions to have the latest version
+     as the SECOND TO LAST ITEM in the list, i.e Always before the version on trunk
+
+     a prerequisite for this test to succeed
```
This line is a bit hard to understand, `a prerequisite .. if ...` but there is no `is what.. ` to me, or is it concatenated from the previous sentence?

guozhangwang commented on tests/kafkatest/tests/streams/streams_test.py:
```diff
@@ -0,0 +1,92 @@
+from kafkatest.tests.kafka_test import KafkaTest
+
+
+class StreamsTest(KafkaTest):
```
nit: Maybe rename to `StreamsTestDriver` to be consistent with other services under `streams.py`.

bbejeck replied:
Since this extends `KafkaTest` and isn't a service itself, what do you think about `StreamsTestBase` or `BaseStreamsTest`?

guozhangwang replied:
Good point, `BaseStreamsTest` is fine.

guozhangwang commented on tests/kafkatest/version.py:
```diff
@@ -91,5 +91,11 @@ def get_version(node=None):
+
+# 1.0.0 versions
+V_1_0_0 = KafkaVersion("1.0.0")
+LATEST_1_0 = V_1_0_0
```
Maybe just add `LATEST_1_1` here as well, though we are not going to use it yet?

bbejeck replied:
Ack

vvcephei replied:
That causes some problems with the dockerfile, since we'd have to download our test jar from the mirror, but it's not there yet. In fact, 1.0.0 is the latest package in the mirror.

mjsax replied:
Those files are uploaded manually for the docker setup. I'll take care of it.

guozhangwang commented on tests/kafkatest/tests/streams/streams_multiple_rolling_upgrade_test.py:
```diff
@@ -0,0 +1,114 @@
+    @parametrize(broker_version=str(LATEST_0_10_2))
+    @parametrize(broker_version=str(LATEST_0_11_0))
+    @parametrize(broker_version=str(LATEST_1_0))
+    @parametrize(broker_version=str(DEV_BRANCH))
```
Could we do `@matrix(broker_version =["...", ...}` as we did in simple benchmark?

bbejeck replied:
ack

bbejeck committed:
Updates per comments, extracted common functionality into base test class

bbejeck commented:
rebased and updated per comments.  

Also since we'd have to wait for this PR for merging to refactor the other tests, I've went ahead and updated the other tests mentioned above to remove the common functionality and extend from the base test class `StreamsTest`

Re-running system tests now.

bbejeck committed:
making call to wrong method

bbejeck committed:
needed to add back old timeout period

bbejeck committed:
renamed base test class to `BaseStreamsTest`

bbejeck committed:
fixed order of topics for assertion, error string format

bbejeck committed:
moved version list to class level, reused in matrix parameter

bbejeck committed:
increase timeout for after broker stop

bbejeck commented:
updated with new system test results.

[standby replica system test result](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1447/)
[broker down system test results](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1449/)
[rolling upgrade system test results](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1450/)

vvcephei commented:
LGTM!

FYI: you'll also want to update the Dockerfile, but I have it updated in my PR, so I reckon it doesn't matter (https://github.com/apache/kafka/pull/4689)
Actually, I forgot the vagrantfile, so between the two of us we have covered both ways to run the tests!

bbejeck commented:
retest this please

mjsax commented on tests/kafkatest/tests/streams/streams_multiple_rolling_upgrade_test.py:
```diff
@@ -0,0 +1,110 @@
+     aws s3api list-objects --bucket kafka-packages --query 'Contents[].{Key:Key}
+    """
+    # adding new version to this list will cover broker and streams version
+    streams_upgrade_versions = [str(LATEST_0_10_2), str(LATEST_0_11_0), str(LATEST_1_0), str(DEV_BRANCH)]
```
what about 0.10.1 and 1.1 ?

bbejeck replied:
Added `0.10.1` I was thinking `0.10.2` was the first version we could do a rolling upgrade from.  I've left out `1.1.0` for the moment as I was thinking this PR will be merged before the 1.1 release.

EDIT the package name of `StreamsSoakTest` is different in 0.10.1  I think @vvcephei  ran into the same issue in https://github.com/apache/kafka/pull/4689. The decision there was to leave out `0.10.1` so IMHO we should do the same here.

mjsax replied:
I see. That makes sense. I run into the same issue and put a workaround in #4636 -- we should clean this up, after we put proper testing for older versions into place as discussed in #4636.

mjsax commented on tests/kafkatest/tests/streams/streams_multiple_rolling_upgrade_test.py:
```diff
@@ -0,0 +1,110 @@
+        self.driver = StreamsSmokeTestDriverService(test_context, self.kafka)
+        self.processor_1 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
+        self.processor_2 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
+        self.processor_3 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
```
Don't we need to set version 0.10.1 for those initially? Otherwise, they will have trunk version

bbejeck replied:
on line 84 of `upgrade_and_verify_start` method, we set the version with the first version from the array and start the node, so they all start with earliest specified.

mjsax reviewed (commented):
Just did a high level review atm.

bbejeck committed:
add 0.10.1 latest version

bbejeck commented:
updated this

bbejeck committed:
removed 0.10.1 version

bbejeck commented:
rebased this

mjsax commented on tests/kafkatest/tests/streams/streams_multiple_rolling_upgrade_test.py:
```diff
@@ -0,0 +1,110 @@
+        self.update_processors_and_verify(self.streams_upgrade_versions)
+
+        # with order reversed now we test downgrading, verification run after each downgrade
+        self.update_processors_and_verify(self.streams_downgrade_versions)
```
Don't we need to stop all processor at the end?

bbejeck replied:
ack

bbejeck committed:
updates per comments, stopping all processors at end of test

bbejeck commented:
updated this and kicked off new system test https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1499/.

EDIT: Tests passed

bbejeck commented:
ping @mjsax  and @guozhangwang for merging

mjsax reviewed (approved):
LGMT.

mjsax commented:
Retest this please

bbejeck commented:
retest this please

bbejeck commented:
retest this please

bbejeck commented:
retest this please

bbejeck commented:
retest this please

bbejeck commented:
@guozhangwang @mjsax 

All tests pass now, call for final review and merging

vvcephei reviewed (approved):
This lgtm.

Your implementation has got me thinking about refactoring the cross-version upgrade/downgrade test to set up the brokers just once and do a bunch of cross-grades like you've done. I think that test would run much faster that way. Something for when I'm on call, maybe...

mjsax commented:
@vvcephei What you suggest will be part of KIP-268 implementation. Compare #4746 and #4758

mjsax commented:
Merged to `trunk`. Thanks for the patch @bbejeck.
